### PR TITLE
[#2199] Implement elisp-lint as an additional CI formatting check

### PR DESCRIPTION
**Pull request by gonewest818**
As per #2199, this pulls in `elisp-lint` as a Cask dependency to check the source code for formatting issues.  The `make lint` target ~runs along with~ now replaces what used to be `make test-checks`. In Travis CI the linter runs during a "check" stage which is positioned prior to the unit tests.

As of the time of this commit ~the source code does not fully pass `make test-lint` but I'm submitting the pull request to show you how things are currently working.~ The most prevalent formatting issue is line widths exceeding the `fill-column`.

`elisp-lint` itself had been dormant up until recently. The package author agreed to pass the repo to me as a maintainer so we can iterate quickly on features if so desired.

-----------------

Not all of the following are relevant.

- [x] The commits are consistent with our [contribution guidelines][1]
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [ ] The new code is not generating bytecode or `M-x checkdoc` warnings
- [ ] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

Thanks!

*If you're just starting out to hack on CIDER you might find this [section of its
manual][2] extremely useful.*

[1]: https://github.com/clojure-emacs/cider/blob/master/.github/CONTRIBUTING.md
[2]: https://cider.readthedocs.io/en/latest/hacking_on_cider/

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -32,13 +32,28 @@ env:
...
-    - EMACS_BINARY=emacs-git-snapshot-travis MAKE_TEST=test-checks
+
+stages:
+  - check
+  - test
+
+jobs:
+  include:
+    # code quality "check" jobs
+    - stage: check
```
Comment (bbatsov) [inline .travis.yml]: The byte-compilation warnings can also be qualified as such a check. 

What's `test-checks` right now? Isn't it just `checkdoc`? Maybe we can give those tasks better names. :-)

**Comment by bbatsov** [comment]
Seems like a good start to me. I wonder if things like `checkdoc` and byte-compilation warnings can't be folded into eldoc - that would simplify the CI setup and would be reusable for pretty much any other Elisp project.

**Comment by gonewest818** [comment]
I guess that sort of restructuring would be available no sooner than Emacs 27, as `eldoc` and `checkdoc` are both part of the distribution. However I don't think I have the time to get pulled into Emacs development, I want to hold the line at package maintenance. :)

File: .travis.yml (hunk 1 above)
Comment (gonewest818) [reply .travis.yml]: Indeed, `test-checks` is "just" `checkdoc` but you've set some configuration in `test/scripts/cider-checks.el` that `elisp-lint` doesn't yet support. So if I were to extend `elisp-lint` to support your configuration needs then `cider-checks.el` can be retired.

**Comment by bbatsov** [comment]
I meant that probably `elint` can invoke them internally and just flag the problems reported by them.

**Comment by gonewest818** [comment]
Ah, that was my misunderstanding.  Yes, elisp-lint already has checkdoc, and a rudimentary ability to byte-compile.

**Comment by gonewest818** [comment]
I told myself "just wrap up the elisp-lint integration, and then move everything to CircleCI", but once again I'm blocked because Travis CI jobs are erroring for no good reason.

**Comment by gonewest818** [comment]
Enabling the `fill-column` check in the linter will require a bunch of reformatting across the code base. The setting is 80 columns. I don't have any problem doing the work per se...  but want to make sure (a) core contributors definitely want long lines to be flagged by the linter, and (b) 80 columns is the desired limit.

**Comment by xiongtx** [comment]
I don't think we should be enforcing the 80-column rule as a matter of principle; certainly not to the extent of failing a build. Sometimes symbol names are long; nothing to do about that.

In any case, Emacs Lisp is not even supposed to wrap at 80 columns--[usually 70 or 72](https://www.gnu.org/software/emacs/manual/html_node/eintr/fill_002dcolumn-Example.html).

**Comment by gonewest818** [comment]
I should have mentioned all linting/validation is advisory only for now. Failures don't block the rest of the build.

The`fill-column` validator is completely disabled. The other validators (byte compile, checkdoc, check-declare, indentation, indent-tabs-mode, trailing whitespace) are running *and* passing on Emacs releases 24, 25, and pre-release 26.  On the emacs.git master branch there is one issue (#2216) arising from the bytecompile check.

**Comment by bbatsov** [comment]
Sorry for the slow response. I think the only real issue we have is with `checkdoc` as it complains about docstrings longer than 80 characters (or maybe that's configurable and I didn't notice that). I don't think we should focus efforts on fixing the line length issues globally, although I always re-flow code longer than 80 chars when I encounter some. After all you have to be able to see two windows side by side fully on a smaller screen. 😉

**Comment by gonewest818** [comment]
I think it's fine to leave the `checkdoc` validation as-is, after all it's widespread in the community and also a validation that MELPA asks for (and enforces? I don't recall.)  Oh, and all this code currently passes checkdoc already.

For the curious, it seems `checkdoc` enforces docstrings need to be < 80 characters wide *and* interesting to know the 80 limit is hard coded. https://github.com/emacs-mirror/emacs/blob/master/lisp/emacs-lisp/checkdoc.el#L1595-L1615

**Comment by bbatsov** [comment]
> I think it's fine to leave the checkdoc validation as-is, after all it's widespread in the community and also a validation that MELPA asks for (and enforces? I don't recall.) Oh, and all this code currently passes checkdoc already.

MELPA requires other things - mostly proper package metadata. It doesn't care about checkdoc at all.

> For the curious, it seems checkdoc enforces docstrings need to be < 80 characters wide and interesting to know the 80 limit is hard coded. https://github.com/emacs-mirror/emacs/blob/master/lisp/emacs-lisp/checkdoc.el#L1595-L1615

Ouch. 😆

**Comment by gonewest818** [comment]
> MELPA requires other things - mostly proper package metadata. It doesn't care about checkdoc at all.

Ok. I was referring to a bullet in the [MELPA contributor guidelines](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.md#expediting-recipe-reviews): "Use checkdoc to make sure that your package follows the conventions for documentation strings."

**Comment by gonewest818** [comment]
Rebased, fixed up a `check-declare` failure, & updated the github pull request template.

The thing I'm not 100% happy about is the `elisp-lint` indentation check. It's not smart enough to understand, e.g. inside `cl-flet` bindings should be indented like functions. You can get around this by setting indent metadata manually in `.dir-locals.el` but I wish it were better.

**Comment by bbatsov** [comment]
It's OK not to get everything perfect from the first try. 😉

**Pull request merged**